zcash_client_sqlite/src/lib.rs

```diff
@@ -474,7 +478,7 @@ impl<P: consensus::Parameters> WalletWrite for WalletDb<rusqlite::Connection, P>
                 }));
 
                 last_scanned_height = Some(block.height());
-                sapling_commitments.extend(block.into_sapling_commitments().into_iter());
+                sapling_commitments.extend(block.into_sapling_commitments().into_iter().map(Some));
```

Comment:
Why do we map this into `Some` and then use `n.take().expect(...)`? Is this just a common idiom that makes it possible to take ownership of parts of an iterator (because `.take()` replaces the iterator's owned chunk with `None`?

Reply:
Yep. `rayon` only provides mutable references to sub-slices, not owned references (which makes sense as you'd otherwise need allocations to split the slice). I don't know if this pattern is common per se for this kind of code, but it's used frequently to implement state machines (e.g. manual `Future` impls), and this is kinda a single-step state machine.